gossip/src/duplicate_shred_handler.rs

```diff
@@ -0,0 +1,451 @@
+use {
+    crate::{
+        cluster_info_duplicate_shred_listener::ClusterInfoDuplicateShredHandler,
+        duplicate_shred::{DuplicateShred, Error},
+    },
+    itertools::Itertools,
+    log::*,
+    solana_ledger::{
+        blockstore::Blockstore, blockstore_meta::DuplicateSlotProof,
+        leader_schedule_cache::LeaderScheduleCache, shred::Shred,
+    },
+    solana_sdk::{clock::Slot, pubkey::Pubkey},
+    std::{
+        collections::{HashMap, HashSet},
+        sync::Arc,
+    },
+};
+
+const CLEANUP_EVERY_N_LOOPS: usize = 10;
+// Normally num_chunks is 3, because there are two shreds (each is one packet)
+// and meta data. So we discard anything larger than 3 chunks.
+const MAX_NUM_CHUNKS: u8 = 3;
+// We only allow each pubkey to send proofs for 5 slots, because normally there
+// is only 1 person sending out duplicate proofs, 1 person is leader for 4 slots,
+// so we allow 5 here to limit the chunk map size.
+const ALLOWED_SLOTS_PER_PUBKEY: usize = 5;
+
+struct ProofChunkMap {
+    num_chunks: u8,
+    missing_chunks: u8,
+    wallclock: u64,
+    chunks: HashMap<u8, Vec<u8>>,
```

Comment:
If max-num-chunks is 3, then instead of `HashMap` better use
```rust
chunks: [Option<Vec<u8>>; 3]
```

Reply:
Tried what you suggested, it asks me to implement copy trait for DuplicateShred because theoretically things can be moved in a vector. The space saved by HashMap to vec shouldn't be that big, because the whole map is ~50MB in the worst case (everyone sending us duplicate proofs for every slot).

Suggestions welcome.